src/sc_client/session.py

```diff
@@ -66,91 +84,110 @@ def _emit_callback(event_id: int, elems: list[int]) -> None:
         event.callback(*[ScAddr(addr) for addr in elems])
 
 
+def _on_open(_) -> None:
+    _ScClientSession.is_open = True
+
+
 def _on_error(_, error: Exception) -> None:
-    close_connection()
-    logger.error(f"{error}")
+    if _ScClientSession.error_handler:
+        _ScClientSession.error_handler(error)
+    else:
+        raise error
 
 
 def _on_close(_, close_status_code, close_msg) -> None:
     close_connection()
-    logger.info(f"{close_status_code}: {close_msg}")
+    _ScClientSession.is_open = False
+    if close_status_code and close_msg:
+        logger.info(f"{close_msg}")
 
 
-def is_connection_established() -> bool:
-    return bool(_ScClientSession.ws_app)
+def set_error_handler(callback) -> None:
+    _ScClientSession.error_handler = callback
 
 
-def set_connection(url: str, do_reconnect, reconnection_time) -> None:
-    _ScClientSession.do_reconnect = do_reconnect
-    _ScClientSession.reconnecion_time = reconnection_time
+def set_reconnect_handler(
+        reconnect_callback, post_reconnect_callback, reconnect_retries: int, reconnect_retry_delay: float) -> None:
+    _ScClientSession.reconnect_callback = reconnect_callback
+    _ScClientSession.post_reconnect_callback = post_reconnect_callback
+    _ScClientSession.reconnect_retries = reconnect_retries
+    _ScClientSession.reconnect_retry_delay = reconnect_retry_delay
 
-    if not is_connection_established():
-        establish_connection(url)
 
-    if _ScClientSession.do_reconnect:
-        set_reconnection(url)
+def set_connection(url: str) -> None:
+    establish_connection(url)
+
+
+def is_connected() -> bool:
+    check_request = json.dumps({"type": "healthcheck"})
+    _ScClientSession.ws_app.send(check_request)
+
+    while not _ScClientSession.last_healthcheck_answer and _ScClientSession.is_open:
+        time.sleep(SERVER_ANSWER_CHECK_TIME)
+
+    result = _ScClientSession.last_healthcheck_answer == "OK"
+    _ScClientSession.last_healthcheck_answer = None
+    return result
 
 
 def establish_connection(url) -> None:
     def run_in_thread():
-        _ScClientSession.ws_app = websocket.WebSocketApp(
-            url,
-            on_message=_on_message,
-            on_error=_on_error,
-            on_close=_on_close,
-        )
-        _ScClientSession.ws_app.run_forever()
+        if not _ScClientSession.ws_app or _ScClientSession.ws_app.url != url:
+            _ScClientSession.ws_app = websocket.WebSocketApp(
+                url,
+                on_message=_on_message,
+                on_error=_on_error,
+                on_close=_on_close,
+            )
+            logger.info(f"Sc-server socket: {_ScClientSession.ws_app.url}")
+        if not _ScClientSession.is_open:
+            try:
+                _on_open(_ScClientSession.ws_app)
```

Comment:
This should be triggered automatically because it should react on the open event of the WebSockets app. We can pass it as an arg to the App [constructor](https://websocket-client.readthedocs.io/en/latest/app.html#websocket._app.WebSocketApp.__init__):

Reply:
No, i tried this way

Reply:
It doesn't work